NLSY/nlsy_imputation.qmd

```diff
@@ -0,0 +1,227 @@
+```
+
+## Multiple Imputation
+
+```{r}
+# Paper that uses multiple imputation: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6049093/
+# Multiple imputation in R reference: https://library.virginia.edu/data/articles/getting-started-with-multiple-imputation-in-r
+init = mice(impute, maxit=10)
+meth = init$method
+predM = init$predictorMatrix
+
+# Set distributions
+num <- c("mom_age") # numerical variables
+log <- c("hisp", "ret_sav", "hown", "both_pars") # binary variables
+poly <- c("mom_educ_hs") # ordered categorical variables
+poly2 <- c("race", "race_eth")# Unordered categorical variable
+
+# Assign Variables to be imputed
+predM[,c("id")] <- 0 # not using as a predictor variable
+predM[c("id"),] <- 0 # not using as a predictor variable
+predM[c("race_eth"),] <- 1 # not using as a predictor variable
+predM[,c("hisp", "mom_age", "ret_sav", "hown", "mom_educ_hs", "both_pars", "race_eth")] <- 0 # not imputing
+meth[c("hisp", "mom_age", "ret_sav", "hown", "mom_educ_hs", "both_pars", "race_eth")]= "" #not imputing missing values, just using as a predictor
+
+# Income and Wealth imputation
+set.seed(25)
+imputed = mice(impute, method=meth, predictorMatrix=predM, m=10)
+
+imputed <- complete(imputed, action = "repeated", include = FALSE)
+
+final <- imputed %>%
+    rename(id = id.1,
+           mom_age = mom_age.1,
+           ret_sav = ret_sav.1,
+           hown = hown.1,
+           mom_educ_hs = mom_educ_hs.1,
+           hisp = hisp.1,
+           both_pars = both_pars.1,
+           race_eth = race_eth.1) %>%
+    select(-starts_with("id."), -starts_with("mom_age."), -starts_with("ret_sav."), -starts_with("hown."), -starts_with("mom_educ_hs."), -starts_with("dad_educ_hs."), -starts_with("both_pars."), -starts_with("hisp."),  -starts_with("race_eth."))
+
+```
+
+```{r}
+# Join back with original income and wealth
+og_values <- impute %>%
+    select(id, pincome, pnetworth)
+
+final <- merge(x=final,y=og_values,
+                    by="id", all.x=TRUE)
+
+income <- final %>%
+    select(id, starts_with("pincome"))
+
+wealth <- final %>%
+    select(id, starts_with("pnetworth"))
+
+pincome <- tidyr::gather(income, key = "imputation", value ="pincome", matches("pincome"), -id)
+pnetworth <- tidyr::gather(wealth, key = "imputation", value ="pnetworth", matches("pnetworth"), -id)
+```
+
+## Distribution Plots for Parent Income (non-imputed and imputed)
+
+```{r}
+ggplot(pincome, aes(x=pincome, fill= imputation)) +
+    geom_density(alpha = 0.5) +
+    facet_wrap(~imputation, scales="free")
+```
+
+## Distribution Plots for Parent Wealth (non-imputed and imputed)
+
+```{r}
+ggplot(pnetworth, aes(x=pnetworth, fill= imputation)) +
+    geom_density(alpha = 0.5) +
+    facet_wrap(~imputation, scales="free")
+```
+
+## Scatter Plots for both parent income and wealth (non-imputed is zero)
+
```

Comment:
The scatter plots reveal some concerning outliers in the imputed sets. The minimum net-worth in the original set is about -$100,000 and some imputed sets have values that approach -$1,000,000. I am not sure what the best way to deal with this is, but one thing we could do is limit min and max values to those from the original set.

It would also be useful to see whisker plots or summary tables with mean, median, min, max, and perhaps 25th and 75th percentile.

Reply:
Added whisker plots! I think this is a great flag. Unfortunately, it does not seem the mice package has this as a built in functionality so we will have to get created. 

I think it would be pretty simple to do this in post-processing, by taking all values that exceed the min and max and replacing them with the data set min and max. However, this doesn't seem to be what we would ideally want - to bound the min and max before the imputation so that the values are generated within those limits. I think we may need to think about what a custom function, with mice embedded into it.